### PR TITLE
Graph Based PP Backward Fix

### PR DESCRIPTION
**Pull request by sanketpurandare**
Fixing backward not being called, gradient scaling and enabling backward with torch.no_grad().
Also overrides UNSHARD, RESHARD and REDUCE_GRAD actions.
Requires: https://github.com/pytorch/pytorch/pull/167277

### CODE REVIEW

File: autoparallel/graph_pp_runner.py [hunk 1]
```diff
@@ -75,6 +79,22 @@ def __init__(
             "unsharded_grads": [],
         }
 
+    def scale_grads(self, grad_scale_factor: int) -> None:
+        """Scale gradients model gradients by `grad_scale_factor`, which should be specified in coordination with the
```
Comment (wconstab) [inline autoparallel/graph_pp_runner.py]: typo, 'gradients' twice

File: autoparallel/graph_pp_runner.py [hunk 2]
```diff
@@ -400,6 +454,19 @@ def __init__(
         schedule: _PipelineScheduleRuntime,
     ):
         self.schedule = schedule
+        if not schedule._backward_requires_autograd:
```
Comment (wconstab) [inline autoparallel/graph_pp_runner.py]: can you say why we need this assertion?  what would happen if we had one graph stage and one non-graph stage?  (not that we want this, but, shouldn't it work fine?

**Review by wconstab — APPROVED** [review]

File: autoparallel/graph_pp_runner.py [hunk 3]
```diff
@@ -243,7 +263,10 @@ def stage_forward(
         composite_args = stage._retrieve_recv_activations(mb_index)
 
     # stage._validate_fwd_input(args, kwargs) Maybe need to validate composite args?
-
+    logger.debug(
```
[the remark this answers is not in the document]
Comment (sanketpurandare) [reply autoparallel/graph_pp_runner.py]: I just wanted to make sure that the code reaches the point that calls the graphs and is not skipped.

File: autoparallel/graph_pp_runner.py (hunk 2 above)
Comment (sanketpurandare) [reply autoparallel/graph_pp_runner.py]: We are running the graph based pp runner in no_grad mode, so the no graph stage will not record any autograd graph for non-graph stages, hence we make this check.

File: autoparallel/graph_pp_runner.py (hunk 1 above)
Comment (sanketpurandare) [reply autoparallel/graph_pp_runner.py]: Fixed!
